esp/services/ws_machine/ws_machineService.cpp

```diff
@@ -3282,3 +3264,222 @@ bool Cws_machineEx::onUpdateComponentStatus(IEspContext &context, IEspUpdateComp
+    Owned<IPropertyTree> dataFolderReq = servicePtr->createDiskUsageReq(envDirectories, "data", thorCluster ? "thor" : "roxie", name);
+    Owned<IPropertyTree> repFolderReq;
+    if (thorCluster)
+        repFolderReq.setown(servicePtr->createDiskUsageReq(envDirectories, "mirror", "thor", name));
+
+    StringBuffer xpath;
+    if (thorCluster)
+        xpath.setf("Software/ThorCluster[@name='%s']/ThorSlaveProcess", name);
+    else
+        xpath.setf("Software/RoxieCluster[@name='%s']/RoxieServerProcess", name);
+    Owned<IPropertyTreeIterator> processes= envRoot->getElements(xpath);
+    ForEach(*processes)
+    {
+        IPropertyTree& process = processes->query();
+        const char* computer = process.queryProp("@computer");
+        if (isEmptyString(computer))
+            throw MakeStringException(ECLWATCH_CANNOT_GET_ENV_INFO, "Failed to get @computer for %s", xpath.str());
+
+        checkAndAddMachineUsageReq(constEnv, computer, logFolderReq, dataFolderReq, repFolderReq, usageReq);
+    }
+
+    if (!thorCluster)
+        return;
+
+    //Read ThorMasterProcess in case it is on a different machine
+    xpath.setf("Software/ThorCluster[@name='%s']/ThorMasterProcess/@computer", name);
+    const char* computer = envRoot->queryProp(xpath);
+    if (isEmptyString(computer))
+        throw MakeStringException(ECLWATCH_CANNOT_GET_ENV_INFO, "Failed to get %s", xpath.str());
+
+    checkAndAddMachineUsageReq(constEnv, computer, logFolderReq, dataFolderReq, repFolderReq, usageReq);
+}
+
+void CUsageCacheReader::checkAndAddMachineUsageReq(IConstEnvironment* constEnv, const char* computer, IPropertyTree* logFolderReq,
+    IPropertyTree* dataFolderReq,  IPropertyTree* repFolderReq, IPropertyTree* usageReq)
+{
+    Owned<IPropertyTree> machineReq = servicePtr->createMachineUsageReq(constEnv, computer);
+    VStringBuffer xpath("Machine[@netAddress='%s']", machineReq->queryProp("@netAddress"));
+    IPropertyTree* foundMachineReqTree = usageReq->queryPropTree(xpath);
+    if (!foundMachineReqTree)
+    {
+        //Not sure we need those folders here. Add them just in case.
+        if (logFolderReq)
+            machineReq->addPropTree(logFolderReq->queryName(), LINK(logFolderReq));
+        if (dataFolderReq)
+            machineReq->addPropTree(dataFolderReq->queryName(), LINK(dataFolderReq));
+        if (repFolderReq)
+            machineReq->addPropTree(repFolderReq->queryName(), LINK(repFolderReq));
+        usageReq->addPropTree(machineReq->queryName(), LINK(machineReq));
+        return;
+    }
+
+    //Add unique disk folders
+    checkAndAddFolderReq(logFolderReq, foundMachineReqTree);
+    checkAndAddFolderReq(dataFolderReq, foundMachineReqTree);
+    checkAndAddFolderReq(repFolderReq, foundMachineReqTree);
+}
+
+void CUsageCacheReader::checkAndAddFolderReq(IPropertyTree* folderReq, IPropertyTree* machineReqTree)
```

Comment:
The code is similar to old code. For example, the addClusterUsageReq() is merged from the readRoxieUsageReq() and the readThorUsageReq(). The checkAndAddMachineUsageReq() is based on the code inside the readRoxieUsageReq() and the readThorUsageReq(), plus machine dedup. The checkAndAddFolderReq() is the old code of adding a folder tree, plus folder dedup. But, the old readRoxieUsageReq() and the readThorUsageReq() cannot be used here because they do not dedup machines and folders.